### PR TITLE
aja: Add audio channel selection to capture

### PR DESCRIPTION
**Pull request by paulh-aja**
### Description

1. Added audio channel selection and FC/LFE channel swapping UI to AJA Source plugin.
1. Adapted audio repacker from DeckLink plugin, for use with 32-bit samples. Code has an ifdef to use SIMD or C version, and currently defaults to use the SIMD version.

The AJA source plugin was originally written to use the `SPEAKERS_7POINT1` layout, with no support for additional layouts. This was done early in development, likely because AJA cards (in the OBS plugin) always process 8ch of audio at a time, regardless of how many channels are present on the wire. After some discussions with another OBS developer, "pkv", I have found that the AJA source plugin will need to implement the rest of the OBS speaker layouts to properly handle capturing a signal containing 2-8ch of audio.

To properly support all of the OBS speaker layouts in the AJA source plugin, we needed to implement an audio "repacker", similar to what the DeckLink capture plugin uses. The audio repacker is used to squash the AJA 8-channel audio buffers down to the sizes required by the various OBS speaker layouts (2 through 8 channels). The AJA plugin could not directly use the audio repacker code from the DeckLink plugin however, because AJA cards always process audio as 32-bit words, whereas the repacker in the DeckLink plugin was hard-coded in SIMD to consider 16-bit samples.

The audio repacker code in this PR has been adapted from the DeckLink audio repacker code, and contains both a new C version and SIMD version. I have tested both versions to ensure they work the same. There are functions to handle the repacking for 32-bit AJA samples, as well as the 16-bit samples from DeckLink. There are also functions to handle the channel swapping behavior found (swap FC and LFE channels in surround layouts) in the DeckLink plugin's audio channel selection.

I've had some discussions with "pkv", the maintainer of the DeckLink audio code, about placing the new repacking code in a more universal place and refactoring both DeckLink and AJA Source plugins to use it. We can discuss that further in this PR. For now the AJA repacker lives "plugins/aja", and the DeckLink repacker remains in "plugins/decklink", which pkv said was acceptable as well.

_NOTE: To support arm64 we will need to use the C version of the repacker, or implement a Neon version of the repacker, if desired._

### Motivation and Context

The AJA source plugin did not properly handle all of the various OBS speaker layouts. The plugin was hard-coded to use `SPEAKERS_7POINT1` by mistake. As best I can tell this caused OBS to treat our audio buffers as 7.1 all of the time, even if the card was capturing fewer than 8ch of audio. The AJA source plugin must always capture a full 8ch buffer from the card, but it needs to be squashed to accommodate the OBS speaker layouts.

### How Has This Been Tested?

Tested by capturing an 8ch audio test signal into an AJA io4K+ card, enabling and disabling individual audio channels in the test signal. I checked all of the speaker layouts in OBS and verified that the correct audio channel(s) from my test signal were heard in OBS on the corresponding channel(s).

### Types of changes

- Bug fix (non-breaking change which fixes an issue)
- New feature (non-breaking change which adds functionality)

### Checklist:

- [x] My code has been run through [clang-format](https://github.com/obsproject/obs-studio/blob/master/.clang-format).
- [x] I have read the [**contributing** document](https://github.com/obsproject/obs-studio/blob/master/CONTRIBUTING.rst).
- [x] My code is not on the master branch.
- [x] The code has been tested.
- [x] All commit messages are properly formatted and commits squashed where appropriate.
- [x] I have included updates to all appropriate documentation.

### CODE REVIEW

**Comment by paulh-aja** [comment]
This PR adapted the audio repacker code from DeckLink plugin, to work with the AJA 32-bit audio buffers. The version of the repacker in this PR has implementations for both 16-bit and 32-bit PCM samples, with the idea that it could eventually be placed into a common location for both plugins to use. This is something that @pkviet and I had been discussion prior to my submission of this PR. However I opted to hold off on that change, as modifications to another plugin should happen in a separate PR. This was also the advice I'd been given in the past from another OBS developer when I had made changes that crossed over into the DeckLink plugin.

That issue aside, I'd like to at least get this PR merged into the 28 beta soon so that I can provide builds to our QA department for testing. It has been difficult for me to keep our internal OBS CI builds up to date with all of the deps and other changes happening in OBS., It is much easier for me to point them to the OBS github CI builds.

**Comment by tt2468** [comment]
I've added Seeking Testers to this PR so that QA can download the CI builds directly from this PR.

**Review by norihiro — COMMENTED** [review]
I guess recent commit d7b363a (PR #6621) was partially reverted by this commit.

File: plugins/aja/aja-source.cpp [hunk 1]
```diff
@@ -901,15 +969,15 @@ static void aja_source_update(void *data, obs_data_t *settings)
...
 		obs_data_get_bool(settings, kUIPropDeactivateWhenNotShowing.id);
+	bool swapFrontCenterLFE =
+		obs_data_get_bool(settings, kUIPropChannelSwap_FC_LFE.id);
 	const std::string &wantCardID =
 		obs_data_get_string(settings, kUIPropDevice.id);
 
-	obs_source_set_async_unbuffered(
-		ajaSource->GetOBSSource(),
-		!obs_data_get_bool(settings, kUIPropBuffering.id));
-
```
[the remark this answers is not in the document]
Comment (pkviet) [reply plugins/aja/aja-source.cpp:912]: please reinstate the previous code

File: plugins/aja/aja-source.cpp [hunk 2]
```diff
@@ -1096,11 +1166,17 @@ static obs_properties_t *aja_source_get_properties(void *data)
...
+	obs_property_t *swap = obs_properties_add_bool(
+		props, kUIPropChannelSwap_FC_LFE.id,
+		obs_module_text(kUIPropChannelSwap_FC_LFE.text));
+	obs_property_set_long_description(swap,
+					  kUIPropChannelSwap_FC_LFE.tooltip);
 	obs_properties_add_bool(
 		props, kUIPropDeactivateWhenNotShowing.id,
 		obs_module_text(kUIPropDeactivateWhenNotShowing.text));
-	obs_properties_add_bool(props, kUIPropBuffering.id,
-				obs_module_text(kUIPropBuffering.text));
```
[the remark this answers is not in the document]
Comment (pkviet) [reply plugins/aja/aja-source.cpp:1103]: please reinstate the previous code

**Review by pkviet — COMMENTED** [review]
Did you check both HDMI and SDI ?
Is the swap required LFE<=> FC required ?
it could be for decklink only.

**Comment by pkviet** [comment]
apart from the points raised by @norihiro , and my minor points, LGTM and i'm in favour of including this in beta

File: plugins/aja/aja-source.cpp (hunk 1 above)
[the remark this answers is not in the document]
Comment (paulh-aja) [reply plugins/aja/aja-source.cpp:912]: Hrm, I thought I had this branch rebased on the very latest OBS master but I guess not. I'll fix it.

File: plugins/aja/audio-repack.c [hunk 3]
```diff
@@ -0,0 +1,268 @@
...
+#ifdef USE_SIMD
+#include <util/sse-intrin.h>
+#endif
+
+#define NUM_CHANNELS 8 /* max until OBS supports higher channel counts */
+
+int check_buffer(struct audio_repack *repack, uint32_t frame_count)
+{
+	const uint32_t new_size =
+		frame_count * repack->base_dst_size + repack->pad_dst_size;
```
[the remark this answers is not in the document]
Comment (paulh-aja) [reply plugins/aja/audio-repack.c:13]: You're thinking of earlier conversations we had when I was first trying to port the repacker code to work w/ the AJA buffers and was encountering crashes. While I was trying to debug that issue I'd tried increasing the size in this `check_buffer` function, which was alleviating the crashes but the real problem ended up being in my modified repack function. Everything is working as it should now though. This `check_buffer` code is the same as what is currently found in the DeckLink repacker code.

File: plugins/aja/aja-source.cpp (hunk 2 above)
[the remark this answers is not in the document]
Comment (paulh-aja) [reply plugins/aja/aja-source.cpp:1103]: Not, not intentional. My branch is definitely rebased on the latest origin/master. It is a bit worrying that your code is not making it into my branch. I'll see if I can figure out what is going on.

File: plugins/aja/aja-source.cpp (hunk 1 above)
[the remark this answers is not in the document]
Comment (paulh-aja) [reply plugins/aja/aja-source.cpp]: I added back the code in question. Must have been a bad merge somewhere along the line in my branch. Sorry for the confusion.

File: plugins/aja/aja-source.cpp (hunk 2 above)
[the remark this answers is not in the document]
Comment (paulh-aja) [reply plugins/aja/aja-source.cpp:1103]: I added back the code in question. Must have been a bad merge somewhere along the line in my branch. Sorry for the confusion.

**Comment by paulh-aja** [comment]
> Did you check both HDMI and SDI ? Is the swap required LFE<=> FC required ? it could be for decklink only.

I haven't specifically tested with HDMI yet, only SDI. Is there a specific concern you had around HDMI capture?

**Comment by paulh-aja** [comment]
> I've added Seeking Testers to this PR so that QA can download the CI builds directly from this PR.

I didn't realize that was all that needed to be done in order to produce CI builds. In that case I will get these builds to our QA department today. I'm fine waiting until the next beta cycle to get this merged.

**Comment by RytoEX** [comment]
> > I've added Seeking Testers to this PR so that QA can download the CI builds directly from this PR.
> 
> I didn't realize that was all that needed to be done in order to produce CI builds. In that case I will get these builds to our QA department today. I'm fine waiting until the next beta cycle to get this merged.

Has this been tested/validated yet?

**Comment by paulh-aja** [comment]
This can be merged. Both the AJA QA lead and I have tested the changes with SDI and HDMI, with all of the speaker layout settings that I've added. I've also tested the FC/LFE swap and that works too. I've verified this by inspecting both the audio meters, monitoring and recording a file to disk with up to 8 channels of audio.

**Pull request merged**